### ISSUE TITLE
The training effect is very poor！

### ISSUE THREAD

**Issue by zygalaxy** [autonomousvision/transfuser#224]
I used four 3080ti, the CPU was 40 cores, and I trained for 41 times, the training duration was 37 hours, and the training code was CUDA_VISIBLE_DEVICES=0,1,2,3 OMP_NUM_THREADS=40 OPENBLAS_NUM_THREADS=1 torchrun --nnodes=1 --nproc_per_node=4 --max_restarts=0 --rdzv_id=1234576890 --rdzv_backend=c10d train.py --root_dir /home/ubuntu/ZY/ZY_2T/PROJECT_AUTO/transfuser/dataset  --parallel_training 1 --batch_size 12, not using all 210G data sets, but only using about 160G data sets, as shown in the following figure.
![image](https://github.com/autonomousvision/transfuser/assets/139957236/831aa484-d999-44dd-88f9-ae68552a51bf)
However, the test result of the trained model is very poor. The test code is, and the visual results are as follows bash leaderboard/scripts/local_evaluation.sh. 
![30](https://github.com/autonomousvision/transfuser/assets/139957236/b7ddea64-bf79-4de6-a653-95141d55345d)
What is the problem? I hope to get a solution, and I will be very grateful.

**Comment by Kait0** [comment]
This looks like a bug or something is wrong with the setup.
Could you set loading to [strict](https://github.com/autonomousvision/transfuser/blob/b602efac325e77f853460ba0d79dc364c2c70488/team_code_transfuser/submission_agent.py#L96) to see whether the model was loaded correctly.
Did the training losses go down properly?

**Comment by zygalaxy** [comment]
I followed your instructions and encountered two issues:
1)My total training loss is shown in the figure below.
![f9feee74ec681f249af53e8488df9f99](https://github.com/autonomousvision/transfuser/assets/139957236/1507d011-0d04-47d1-8df4-d017ad42f585)
2)The error message when strict is set to True is different for my model and the Transfuser model (both fail to execute with strict set to True). Figure 1 shows the error message for my_train_model, and Figure 2 shows the error message for the Transfuser model (since the full message cannot be displayed, I have also submitted error.txt for my_model).
![image](https://github.com/autonomousvision/transfuser/assets/139957236/bfa49fd1-b620-411a-99f9-106948e87bc1)
![image](https://github.com/autonomousvision/transfuser/assets/139957236/eb214da7-4776-4308-a367-80a499acbdeb)
[error.txt](https://github.com/user-attachments/files/15894337/error.txt)

The loss seems normal, so as you mentioned, the issue is likely related to model loading. However, based on the error messages, I can't determine the exact problem. I hope to get your advice, thank you!

**Comment by Kait0** [comment]
The error for the Transfuser model is normal, it was trained with an older codebase.

It probably has something to do with the name.
Did you change how the model was saved [here](https://github.com/autonomousvision/transfuser/blob/b602efac325e77f853460ba0d79dc364c2c70488/team_code_transfuser/train.py#L383)?

When trained with distributed data parallel we remove part of the name that comes from ddp [here](https://github.com/autonomousvision/transfuser/blob/b602efac325e77f853460ba0d79dc364c2c70488/team_code_transfuser/submission_agent.py#L95).

Are there any optimizer files in your model folder (that could get loaded accidentally)?

**Comment by zygalaxy** [comment]
You are right. I found that my training results seem to contain optimization files (as shown in Figure 1 below). My submission_agent.py is shown in Figure 2 below, and the training code is CUDA_VISIBLE_DEVICES=0,1,2. 3 OMP_NUM_THREADS=40 OPENBLAS_NUM_THREADS=1 torchrun --nnodes=1 --nproc_per_node=4 --max_restarts=0 --rdzv_id=1234576890 --rdzv_backend=c10d train.py - -root _ dir/home/Ubuntu/zy/zy _ 2t/project _ auto/transfuser/dataset-parallel _ training1-batch _ size12, what's wrong with this training result? When I use model_41.pth+optimizer_41.pth, the problem is that the vehicle directly turns right and deviates from the route. When I only use the model_41.pth file, the problem is that the vehicle does not move at all (if the transfuser model is used, the vehicle runs well).
![image](https://github.com/autonomousvision/transfuser/assets/139957236/ce601322-68da-47a4-a168-3b14118baf4d)
![113b9f5be50f98be2791a5411749a69d](https://github.com/autonomousvision/transfuser/assets/139957236/3e25a21d-6e8a-4ff4-ac5f-7f968bdf7ccd)
![image](https://github.com/autonomousvision/transfuser/assets/139957236/f96f0d13-cf95-4889-97b3-9f8158dd5de1)

**Comment by Kait0** [comment]
Your training result seems correct.
Inference needs to be different. The correct way to use the code is to copy the args.txt + model_31.pth file into a new folder, my_model in your case, (no optimizer files) and use this folder for inference.
Then you should not get any error when you load the model with strict.
Can you send the visualization file again when you do this.

The inference code loads every .pth file in the given folder and ensembles the models.

**Comment by zygalaxy** [comment]
Thank you very much for your help. It seems that my model has always been good, but it can't run normally because of the existence of the optimization file. In fact, I found that both model_31.pth and model_41.pth can be executed normally, but it takes a long time for the vehicle to start, which seems to be too cautious? Now I have multiple models, such as 31, 35, 41, etc. How should I combine them to get the best results?

**Comment by Kait0** [comment]
Typically epoch 31 yields the best or close to the best results. 
Sometimes TransFuser gets stuck due to causal confusion. If it does not happen very often this is normal.
After a while a crepping mechanism will unblock it if this happens. This is discussed in Table 10 of the paper.

**Comment by zygalaxy** [comment]
Then it means that I should directly change the default epoch=41 to 31 and use 31 as the final model, which can also reduce the training time.

**Comment by Kait0** [comment]
you can do that. We have done that as well in our follow up work.

**Comment by zygalaxy** [comment]
Ok, thank you for your patience. I appreciate it very much. I will close this question.
